File: 2024/day2.py
def report_is_safe(report):
    diffs = [x - y for x, y in zip(report[:-1], report[1:])]
    return all(1 <= d <= 3 for d in diffs) or all(-3 <= d <= -1 for d in diffs)

def dampened_report_is_safe(report):
    return any(report_is_safe(report[:i] + report[i+1:]) for i in range(len(report)))

with open("day2.in", "r") as f:
    reports = [[int(level) for level in report.split()] for report in f]
    print(sum(1 for report in reports if report_is_safe(report)))
    print(sum(1 for report in reports if dampened_report_is_safe(report)))

File: 2024/day2.go
package main

import "strings"
import "strconv"
import "bufio"
import "os"

func report_is_safe(report []int) bool {
	prev := report[0]
	if report[1] > report[0] {
		for _, level := range report[1:] {
			if level - prev > 3 || level - prev < 1 {
				return false
			}
			prev = level
		}
	} else {
		for _, level := range report[1:] {
			if prev - level > 3 || prev - level < 1 {
				return false
			}
			prev = level
		}
	}
	return true
}

func dampened_report_is_safe(report []int) bool {
	for i := range report {
		if report_is_safe(append(append([]int{}, report[:i]...), report[i+1:]...)) {
			return true
		}
	}
	return false
}

func day2(path string) (answer1, answer2 int, err error) {
	file, err := os.Open(path)
	if err != nil {
		return
	}
	defer file.Close()
	lines := bufio.NewScanner(file)
	for lines.Scan() {
		words := bufio.NewScanner(strings.NewReader(lines.Text()))
		words.Split(bufio.ScanWords)
		var report []int
		for words.Scan() {
			var num int
			num, err = strconv.Atoi(words.Text())
			if err != nil {
				return
			}
			report = append(report, num)
		}
		if report_is_safe(report) {
			answer1 += 1
		}
		if dampened_report_is_safe(report) {
			answer2 += 1
		}
	}
	return
}
